classes/controller/FrontController.php

```diff
@@ -1984,7 +1984,34 @@ protected function getAlternativeLangsUrl()
         }
 
         foreach ($languages as $lang) {
-            $alternativeLangs[$lang['language_code']] = $this->context->link->getLanguageLink($lang['id_lang']);
+            $langUrl = $this->context->link->getLanguageLink($lang['id_lang']);
+
+            $params = [];
+            $url_details = parse_url($langUrl);
+
+            if (!empty($url_details['query'])) {
+                parse_str($url_details['query'], $query);
+                foreach ($query as $key => $value) {
+                    $params[Tools::safeOutput($key)] = Tools::safeOutput($value);
+                }
+            }
+
+            $excluded_key = ['isolang', 'id_lang', 'controller', 'fc', 'id_product', 'id_category', 'id_manufacturer', 'id_supplier', 'id_cms'];
+            $excluded_key = array_merge($excluded_key, $this->redirectionExtraExcludedKeys);
+            foreach ($_GET as $key => $value) {
+                if (!in_array($key, $excluded_key) && Validate::isUrl($key) && Validate::isUrl($value)) {
+                    $params[Tools::safeOutput($key)] = Tools::safeOutput($value);
+                }
+            }
+
+            $str_params = http_build_query($params, '', '&');
+            if (!empty($str_params)) {
+                $langUrl = preg_replace('/^([^?]*)?.*$/', '$1', $langUrl) . '?' . $str_params;
+            } else {
+                $langUrl = preg_replace('/^([^?]*)?.*$/', '$1', $langUrl);
+            }
```

Comment:
I think those lines are the same as [these lines](https://github.com/PrestaShop/PrestaShop/blob/develop/classes/controller/FrontController.php#L800-L822), you can maybe extract that logic to reuse it in both methods?

Reply:
Yes, I think so, too, but I just don't know how to call it.

A `protected` in this controller sounds good to you? Who'd get called twice, once for canonical, once here.

Reply:
Ping @atomiix

Reply:
@Sinepel I think you can go `private` with this new method. And yes, It will be called twice like you said ;)

Reply:
@atomiix I just pushed the fix :)

Reply:
thanks @Sinepel ! you just need to run cs fixer on your files so that the CI is happy and you'll be all set ;)